File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if(root == null){
            return null;
        }
          if(containsNodeInSubtree(root, p) && containsNodeInSubtree(root, q)){
     
        TreeNode left = lowestCommonAncestor(root.left, p, q); 
        TreeNode right = lowestCommonAncestor(root.right, p, q); 

        if(left == null && right == null){
            return root;
        }

        if(left != null){
            return left; 
        } else {
            return right;
        }
          }

        return null;
    }


    public boolean containsNodeInSubtree(TreeNode root, TreeNode numToSearchFor){
       if(numToSearchFor == null){
        return true;
       }
       if(root == null){
        return false; 
       }

       if(root.val == numToSearchFor.val){
        return true; 
       } 

       return containsNodeInSubtree(root.left, numToSearchFor) || containsNodeInSubtree(root.right, numToSearchFor); 
    }
}

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if(not root): 
            return [] 

        queue = deque() 
        queue.append(root)
        res = [] 

        while(queue): 
            toAdd = queue[-1]   
            numNodes = len(queue) 

            for i in range(numNodes): 
                curr = queue.popleft()

                if(curr.left): 
                    queue.append(curr.left)

                if(curr.right):
                    queue.append(curr.right)

            res.append(toAdd.val)

        return res 